### ISSUE TITLE
Incorrect output at 1024x1024 resolution

### ISSUE THREAD

**Issue by developer0hye** [developer0hye/candle-birefnet#1]
## Problem

Inference produces incorrect (noisy) segmentation masks at 1024x1024 resolution, while 384x384 works correctly and matches PyTorch output within 6.87e-5 max error.

## Steps to Reproduce

```bash
# Works correctly
cargo run --example inference --release -- --image examples/Helicopter.jpg --size 384

# Produces garbage output
cargo run --example inference --release -- --image examples/Helicopter.jpg --size 1024
```

## Expected Behavior

Output should match PyTorch BiRefNet at any resolution. PyTorch produces correct results at 1024x1024.

## Possible Causes

- Swin backbone window padding/attention mask computation at non-standard spatial dimensions
  - 1024/4 = 256, and 256 is not divisible by window_size=12 (256/12 = 21.33)
  - Padding and attention mask logic may have edge cases at these dimensions
- Bilinear interpolation `align_corners=True` behavior differences at larger scales
- `image2patches` grid computation when spatial dims don't divide evenly

## Environment

- candle-core 0.9.2
- candle-swin (Swin-V1-Large, window_size=12)
- BiRefNet default config (Swin-L + ASPPDeformable)

**Comment by developer0hye** [comment]
## Root Cause Analysis

After stage-by-stage debugging, the issue has been isolated:

### What works correctly
- Swin-L backbone at 1024x1024: all 4 stages match PyTorch within 1.6e-3
- Encoder (multi-scale cat + context augmentation): OK
- Squeeze module: OK (max_diff=2.25e-4)
- Stage 4 decoder (BasicDecBlk + GDT attention): OK (max_diff=4.76e-5)
- `image2patches` at all grid sizes: exact match (diff=0)

### Where it diverges
`ipt_blk4` (SimpleConvs) at stage 3 produces different output despite **identical input** (patches3, diff=0) and **identical weights** (conv1.weight, diff=0).

- Candle ipt_feat3 abs_mean: **0.271591**
- PyTorch ipt_feat3 abs_mean: **0.267664**
- Difference: ~1.5%

### Cause
This is a **candle-core Conv2d numerical precision issue** when processing tensors with large channel counts (768 input channels) at 64x64 spatial resolution. The accumulation order in candle's im2col + matmul conv2d implementation differs from PyTorch's, causing fp32 rounding differences that are larger than typical (~1e-6) rounding.

This small per-layer error (~1.5%) compounds through the 4 decoder stages, resulting in the final output divergence (max_diff=16.7).

### Note
- At 384x384, the spatial dims are smaller and channel counts happen to accumulate more favorably, keeping errors within 6.87e-5.
- This is NOT a bug in candle-birefnet's architecture or weight loading — it's a precision characteristic of candle-core's conv2d kernel.

**Comment by developer0hye** [comment]
## Resolved

Root cause identified: **candle-core `conv2d_tiled` bug** ([huggingface/candle#3404](https://github.com/huggingface/candle/issues/3404)).

### Fix
Switching candle-core's `DEFAULT_CONV2D_IMPL` from `TiledIm2Col` to `FullIm2Col` resolves the issue completely.

| Resolution | Before fix | After fix |
|-----------|-----------|-----------|
| 384x384 | 6.87e-5 | 6.87e-5 |
| 1024x1024 | **16.7** | **1.63e-4** |

### Workaround applied
Until candle upstream fixes the bug, this project uses a patched candle-core via path dependency that forces `FullIm2Col` for Conv2d.

Tracking: huggingface/candle#3404

**Comment by developer0hye** [comment]
sovled
